firedrake/projection.py

```diff
@@ -35,72 +81,136 @@ def project(v, V, bcs=None, mesh=None,
+    val.rename(name)
+    return val
 
-    if isinstance(V, functionspaceimpl.WithGeometry):
-        ret = function.Function(V, name=name)
-    elif isinstance(V, function.Function):
-        ret = V
-        V = V.function_space()
-    else:
-        raise RuntimeError(
-            'Can only project into functions and function spaces, not %r'
-            % type(V))
 
-    if isinstance(v, expression.Expression):
-        shape = v.value_shape()
-        # Build a function space that supports PointEvaluation so that
-        # we can interpolate into it.
-        if isinstance(V.ufl_element().degree(), tuple):
-            deg = max(V.ufl_element().degree())
-        else:
-            deg = V.ufl_element().degree()
+class Assigner(object):
+    def __init__(self, source, target):
+        self.source = source
+        self.target = target
 
-        if v.rank() == 0:
-            fs = functionspace.FunctionSpace(V.mesh(), 'DG', deg+1)
-        elif v.rank() == 1:
-            fs = functionspace.VectorFunctionSpace(V.mesh(), 'DG',
-                                                   deg+1,
-                                                   dim=shape[0])
+    def project(self):
+        self.target.assign(self.source)
+        return self.target
+
+
+class ProjectorBase(object, metaclass=abc.ABCMeta):
+    def __init__(self, source, target, bcs=None, solver_parameters=None,
+                 form_compiler_parameters=None, constant_jacobian=True):
+        if solver_parameters is None:
+            solver_parameters = {}
         else:
-            fs = functionspace.TensorFunctionSpace(V.mesh(), 'DG',
-                                                   deg+1,
-                                                   shape=shape)
-        f = function.Function(fs)
-        f.interpolate(v)
-        v = f
-    elif isinstance(v, function.Function):
-        if v.function_space().mesh() != ret.function_space().mesh():
-            raise RuntimeError("Can't project between mismatching meshes")
-    elif not isinstance(v, ufl.core.expr.Expr):
-        raise RuntimeError("Can only project from expressions and functions, not %r" % type(v))
-
-    if v.ufl_shape != ret.ufl_shape:
-        raise RuntimeError('Shape mismatch between source %s and target function spaces %s in project' %
-                           (v.ufl_shape, ret.ufl_shape))
-
-    p = ufl_expr.TestFunction(V)
-    q = ufl_expr.TrialFunction(V)
-    a = ufl.inner(p, q) * ufl.dx(domain=V.mesh())
-    L = ufl.inner(p, v) * ufl.dx(domain=V.mesh())
-
-    # Default to 1e-8 relative tolerance
-    if solver_parameters is None:
-        solver_parameters = {'ksp_type': 'cg', 'ksp_rtol': 1e-8}
-    else:
-        solver_parameters.setdefault('ksp_type', 'cg')
-        solver_parameters.setdefault('ksp_rtol', 1e-8)
+            solver_parameters = solver_parameters.copy()
+        solver_parameters.setdefault("ksp_type", "cg")
```

Comment:
To maintain compatibility, need to also say:
`solver_parameters.setdefault("ksp_rtol", 1e-8)`